### ISSUE TITLE
Sort top-level declarations topologically

### ISSUE THREAD

**Issue by apaszke** [google-research/dex-lang#409]
It is quite annoying that all the helper functions have to always go before the place where they're used. Pretty much every language has some support for out-of-order definitions (e.g. Haskell does this automatically, C has forward declarations).

**Comment by oxinabox** [comment]
There is a fun intestection between this, and #410  and #412.
You need a dependency graph for this; which is the core requirement for reactive programming.
Invalidating compilation catches is a super typical application of reactive programming to compilation,
and invalidating output (#410) is that same application to but exectution rather than compilation.
Interactivity (#410) basically is the same thing also, invalidate  and recompute the parts of the execution that depends on the interactive value that has been mutated.
(Though since Dex is pure that simplifies it somewhat, at least if we are willing to recompute things that haven't changed)

**Comment by apaszke** [comment]
Yeah, there's definitely a lot of connections between those. It would be awesome if we could just share a single piece of compiler infrastructure for it all. We were pondering about having a central cache that would trigger compilation/evaluation, etc. and so that piece would probably also be a good place to answer queries for dependencies between different pieces of code.
